compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs

```diff
@@ -450,27 +450,35 @@ generateDocTestModuleRule =
-      allFiles <- listFilesRecursive pkgDb
-      let dalfs = filter ((== ".dalf") . takeExtension) allFiles
-      forM dalfs $ \dalf -> do
-          dalfPkgOrErr <- readDalfPackage dalf
-          let baseName = takeBaseName dalf
-          let getUnitId pkg
-          -- If we use data-dependencies we can end up with multiple DALFs for daml-prim/daml-stdlib
-          -- one per version. The one shipped with the SDK is called daml-prim.dalf and daml-stdlib-$VERSION.dalf
-          -- and have the same unit ids, so we do not need to strip package ids.
-          -- The one coming from daml-prim will be called daml-prim-$PKGID.dalf daml-stdlib-$PKGID.dalf
-          -- To avoid collisions, we include this hash in the unit id so we also don’t want to strip
-          -- package ids here.
-                  | "daml-prim" `isPrefixOf` baseName = (stringToUnitId baseName, pkg)
-                  | "daml-stdlib" `isPrefixOf` baseName = (stringToUnitId baseName, pkg)
-                  | otherwise = ( stringToUnitId $ parseUnitId baseName $ LF.dalfPackageId pkg, pkg)
-          pure (fmap getUnitId dalfPkgOrErr)
-  return (diags, Map.fromList pkgs)
+    versionedPackageDbs <- getPackageDbs version userPkgDbs
+    (diags, pkgs) <-
+        fmap (partitionEithers . concat) $
+        forM versionedPackageDbs $ \pkgDb -> do
+            allFiles <- listFilesRecursive pkgDb
+            let dalfs = filter ((== ".dalf") . takeExtension) allFiles
+            forM dalfs $ \dalf -> do
+                dalfPkgOrErr <- readDalfPackage dalf
+                let baseName = takeBaseName dalf
+                let getUnitId pkg
+                -- If we use data-dependencies we can end up with multiple DALFs for daml-prim/daml-stdlib
+                -- one per version. The one shipped with the SDK is called daml-prim.dalf and daml-stdlib-$VERSION.dalf
+                -- and have the same unit ids, so we do not need to strip package ids.
+                -- The one coming from daml-prim will be called daml-prim-$PKGID.dalf daml-stdlib-$PKGID.dalf
+                -- To avoid collisions, we include this hash in the unit id so we also don’t want to strip
+                -- package ids here.
+                        | "daml-prim" `isPrefixOf` baseName = (stringToUnitId baseName, pkg)
+                        | "daml-stdlib" `isPrefixOf` baseName = (stringToUnitId baseName, pkg)
+                        | otherwise = ( stringToUnitId $ parseUnitId baseName $ LF.dalfPackageId pkg, pkg)
+                pure (fmap getUnitId dalfPkgOrErr)
+
+    let unitIdConflicts = Map.filter ((>=2) . Set.size) . Map.fromListWith Set.union $
+            [ (unitId, Set.singleton (LF.dalfPackageId dalfPkg))
+            | (unitId, dalfPkg) <- pkgs ]
+    when (not $ Map.null unitIdConflicts) $ do
+        fail $ "Transitive dependencies with same unit id but conflicting package ids: "
+            ++ intercalate ", " (map show $ Map.keys unitIdConflicts)
```

Comment:
Should we include the the package ids in the error? Seems potentially useful for debugging.

Reply:
Sounds good, added